Fix dark mode issue and layout tweaks

viniciussoares opened the pull request:
Hi. Recently I've started using this app and thought I could help a little. So I did the following:

- Need list layout. Modern look and improved UX.
- Empty State.
- Fixed dark mode issue.
- Respect safe area.

I didn't found any contributing guidelines. So if there's any issues, let me know :)

Screenshots:




| This PR | Current Version | 
|:-------------------------:|:-------------------------:|
|<img alt="Portrait" width="500" src="https://user-images.githubusercontent.com/12788847/84182003-1985c180-aa60-11ea-805e-ba118e8546b5.png">  |  <img alt="Current Version" src="https://user-images.githubusercontent.com/12788847/84182768-4b4b5800-aa61-11ea-8321-22b94253adeb.png">|

|Splash | Empty State | 
|:-------------------------:|:-------------------------:|
|<img alt="Splash" src="https://user-images.githubusercontent.com/12788847/84181809-d75c8000-aa5f-11ea-88e6-ce6b0b6779d8.png">  |  <img alt="Empty State" src="https://user-images.githubusercontent.com/12788847/84181811-d88dad00-aa5f-11ea-8fa2-b9ee5b94c140.png">|

| Portrait | Portrait 2 | 
|:-------------------------:|:-------------------------:|
|<img alt="Portrait" src="https://user-images.githubusercontent.com/12788847/84182003-1985c180-aa60-11ea-805e-ba118e8546b5.png">  |  <img alt="Portrait 2" src="https://user-images.githubusercontent.com/12788847/84182065-34583600-aa60-11ea-82b6-045a27944df4.png">|


| Landscape |
|:-------------------------:|
| <img alt="Landscape 1" src="https://user-images.githubusercontent.com/12788847/84182195-62d61100-aa60-11ea-8365-f5b9cb24bc8c.png"> |
| <img alt="Landscape 2" src="https://user-images.githubusercontent.com/12788847/84182217-6bc6e280-aa60-11ea-948f-0443c41ac69f.png"> |

npmccallum commented:
@viniciussoares There's a lot of good stuff here. A cursory glance looks like it is mostly pretty good. However, I'll let @justin-stephenson take a closer look.

One thing I would like, and I honestly don't remember the present situation, is for the token code to be scaled as large as possible without wrapping. This improves readability.

viniciussoares commented:
@npmccallum make sense. Now the codeLabel scales to fit the parent width :)

| |
|:-------------------------:|
| <img width="300" alt="iPad" src="https://user-images.githubusercontent.com/12788847/84184797-5e135c00-aa64-11ea-8c9a-57ed100ae1cc.png"> | 
| <img width="300" alt="iPad Landscape" src="https://user-images.githubusercontent.com/12788847/84184827-6a97b480-aa64-11ea-968d-3f4815fa41be.png"> |
| <img width="300" alt="iPhone X" src="https://user-images.githubusercontent.com/12788847/84184862-7daa8480-aa64-11ea-97a8-0c3870b355a3.png"> |

justin-stephenson commented:
@viniciussoares This looks great! I will spend some time testing this on my iOS devices and make sure it does not cause any AutoLayout problems.

npmccallum commented:
@viniciussoares A little bit more horizontal margin/padding on the token field would be good. But otherwise, looks great.

Two more questions:

1. What does the locked icon look like? I don't see it in any of the pictures.

2. Which side is the icon on when left-to-right languages are used?

viniciussoares commented:
@npmccallum I forgot to add support for RTL. Here's the screenshots:

| LTR | RTL | 
|:-------------------------:|:-------------------------:|
|<img alt="LTR" src="https://user-images.githubusercontent.com/12788847/84193858-7ab69080-aa72-11ea-8822-be24fa8b226d.png">  |  <img alt="RTL" src="https://user-images.githubusercontent.com/12788847/84193873-82763500-aa72-11ea-94dd-fabba646445a.png">|

justin-stephenson commented:
@viniciussoares I don't like to force light mode on users, I believe we need to make some changes here to ensure we have a user interface which adapts to Light or Dark mode appropriately. This is a great start however.

One other comment is that the token label/account text is quite small, it is noticeable when I run this PR code on my iOS device.

viniciussoares commented:
> I don't like to force light mode on users

Agree. I was planning to add support do dark mode, that's why i created that Colors extension. I can add that to this PR if you want to :)

> One other comment is that the token label/account text is quite small,

I'll look into it.

justin-stephenson commented:
Thank you @viniciussoares - Please upload the dark mode changes to this PR when they are ready.

The only issue I see currently in my testing is mishandling token codes with more than 6 digits.

![9214651A-AF32-4689-B5CB-51902B55DA49_4_5005_c](https://user-images.githubusercontent.com/13416969/84427829-8aa9ae00-abf3-11ea-95bc-469afac16869.jpeg)

viniciussoares commented:
@justin-stephenson i've added dark mode and fixed the issue with codes that have more than 6 digits :)

_Bonus: on iOS 13 devices i've replaced the old collectionview gestures with context menu (Only for iOS 13 or above)._

| Dark Mode | iOS 13 Context Menu | 
|:-------------------------:|:-------------------------:|
|<img alt="Dark Mode" src="https://user-images.githubusercontent.com/12788847/84577919-cfba1580-ad96-11ea-9bc8-c25af337aa24.png">  |  <img alt="Context Menu" src="https://user-images.githubusercontent.com/12788847/84577922-d47ec980-ad96-11ea-8489-55d01dba477b.png">|

justin-stephenson commented:
Much appreciated @viniciussoares - I will review this in more detail over the next few days.

npmccallum commented:
@viniciussoares @justin-stephenson How much do we care about supporting old iOS versions on new releases? My understanding is that Apple users are very commonly on the latest release (unlike Android).

justin-stephenson commented:
@viniciussoares I definitely prefer the iOS 13+ Context menu for the press + hold gesture, but this removes the re-ordering tokens functionality. I would like to emulate the iOS 13+ home screen App long gesture behavior to implement both reordering and the pop-up context menu:

  - Initial press + hold displays the context menu
  - After the minimum time, press + hold with finger movement causes the re-ordering functionality to trigger. i.e. the gesture recognizer moves to the `UIGestureRecognizer.State.changed` state

I would also like to request confirmation when selecting `Delete` from the context menu, similar to the Home menu `Delete app` confirmation.

Thanks again for your work on this!

justin-stephenson commented:
@npmccallum From initial research iOS 13 is supported and provided for iPhone 6S(2015 release) and later devices, iPhone 6(2014) and older would not be supported. Based on this I don't see a problem with supporting only iOS 13 or higher, but i'm open to input from users.

This PR allows the context menu pop-up feature as shown in the screenshots here for iOS 13 but breaks the swipe gesture token removal. I am not sure if we want or need to keep the swipe gesture removal.

viniciussoares commented:
For now i will remove the context menu, and then this can be discussed elsewhere. Did a rebase as well.

justin-stephenson commented on FreeOTP/FRNavigationController.swift:
```diff
@@ -0,0 +1,33 @@
+
+import UIKit
+
+class FRNavigationController: UINavigationController {
```
What is the meaning of the naming choice `FRNavigationController` ?

viniciussoares replied:
The first two letter of the project name. I wasn't very inspired that day haha

justin-stephenson replied:
I suppose `MainNavigationController` might be better, or something similar

justin-stephenson commented on FreeOTP.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,8 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		2489072124910C89001C75A9 /* UIFont+Ext.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2489072024910C89001C75A9 /* UIFont+Ext.swift */; };
```
I think it would be nice to move `Color.swift` and `UIFont+Ext.swift` into a file named more generically, like `Style.swift`, containing the code from both these files.

viniciussoares replied:
Sure :)

justin-stephenson commented on FreeOTP.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,8 @@
 
 /* Begin PBXBuildFile section */
+		2489072124910C89001C75A9 /* UIFont+Ext.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2489072024910C89001C75A9 /* UIFont+Ext.swift */; };
+		2489072324910D3B001C75A9 /* Screen.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2489072224910D3A001C75A9 /* Screen.swift */; };
```
For reasons similar to above, can we rename `Screen.swift` to `Device.swift` which will be used to handle device specific changes/code like dynamic screen size.

justin-stephenson commented on FreeOTP/RTLSupport.swift:
```diff
@@ -0,0 +1,36 @@
+    func rtlRightToLeft(of constrainable: Constrainable, offset: CGFloat = 0) -> Constraint {
+        usingDefaultLayoutDirection ? rightToLeft(of: constrainable, offset: offset) : leftToRight(of: constrainable, offset: -offset)
+    }
+}
```
This code can be moved into `TokensViewController.swift`, I don't think it needs to be in its own individual file.

justin-stephenson replied:
or `TokenCell.swift`

viniciussoares replied:
I think this may be used elsewhere if other views start to be made using view code (Which seems a trend in the community).

Note that this extension is only required because the app still supports iOS 9.

justin-stephenson commented on FreeOTP/UICollectionViewFlowLayout.swift:
```diff
@@ -22,9 +22,22 @@ import Foundation
 import UIKit
 
 extension UICollectionViewFlowLayout {
+    private var isLandscace: Bool {
```
Should be `isLandscape` ?

justin-stephenson commented on FreeOTP/UICollectionViewFlowLayout.swift:
```diff
@@ -22,9 +22,22 @@ import Foundation
     func columnWidth(_ collectionView: UICollectionView, numCols: CGFloat) -> CGFloat {
+        var width = collectionView.frame.size.width
+
+        if #available(iOS 11.0, *), isLandscace {
```
Here as well

justin-stephenson commented on FreeOTP/Info.plist:
```diff
@@ -2,6 +2,8 @@
 <plist version="1.0">
 <dict>
+	<key>UIUserInterfaceStyle</key>
+	<string>Light</string>
```
This should be removed I believe, to allow the App to choose the device configured UI Style light or dark.

justin-stephenson commented:
> Sorry about the delay, it's been a busy week.

No problem!

> 
> For now i will remove the context menu, and then this can be discussed elsewhere. Did a rebase as well.

Okay agreed, we can continue that discussion in a separate PR. We would like to take advantage of iOS 13 features like context menus, we do need to keep supporting older iOS versions(for now) and make sure we those users do not lose any functionality.
> 
> Non-related subject, why do the option to add a secret manually was removed? This change make it impossible to add an account created from the device itself. :/

This was done in an effort to simplify and streamline the user experience. Note that `otpauth://` URL links opened on the device will forward to FreeOTP, this is a tested use case and invokes the scan token navigation wizard as you will see in the App Delegate.

I added some review comments in-line, most are minor and overall the code looks good to me. I would also like to ask if you can please
- Address these review items
- Squash the commits into only a few logically grouped commits
- Cleanup the git history

Once done, I think this PR will be ready to be merged, unless Nathaniel may have other comments.

The actual App testing I have done shows no problems and makes the UI look much cleaner and modern. Thanks very much for this!

viniciussoares commented:
✅

npmccallum commented:
@justin-stephenson LGTM Thanks @viniciussoares !

justin-stephenson commented:
Thanks a lot @viniciussoares !
